### PR TITLE
Fix request signing middleware

### LINKED ISSUE #2779: Request signing middleware does not strip listen path
**Branch/Environment/Version**
- Branch/Version: master/release-2.9

**Describe the bug**
When we sign a request to the upstream, we must use the exact final path/method. 
However, right now, if Tyk API has listen path, it does not get stripped out, so request signature is generated with "ListenPath + Path", while it should be "Path" (only if "Strip listen path" is turned on).

Currently, there is a workaround if you are using URL rewrite, it uses the right path. And current tests are passing only because they use URL rewrite.

**Reproduction steps**
Create API with non-empty listen path, enable request signing (no matter if HMAC or RSA). Sent request to the upstream, and see gateway debug logs for  "Generated sig string: ..." message. Signature in the gateway will contain listen path, even if listen path stripping is enabled.

### PR DESCRIPTION
**Pull request by komalsukhani**
Fixes #2779 

1. Handle case of url-rewrite and strip listen path while generating signature string
2. Adds additional checks for input
3. Added more test cases

### CODE REVIEW

File: gateway/mw_request_signing.go [hunk 1]
```diff
@@ -101,7 +101,12 @@ func (s *RequestSigning) ProcessRequest(w http.ResponseWriter, r *http.Request,
 	}
 
 	headers := generateHeaderList(r, s.Spec.RequestSigning.HeaderList)
-	signatureString, err := generateHMACSignatureStringFromRequest(r, headers)
+
+	path := r.URL.Path
+	if s.Spec.Proxy.StripListenPath {
+		path = s.Spec.StripListenPath(r, r.URL.Path)
+	}
+	signatureString, err := generateHMACSignatureStringFromRequest(r, headers, path)
```
Comment (buger) [inline gateway/mw_request_signing.go]: Do not think it is only about stripping listen path. What if user used URL rewrite, and updated the path, will your change pick this up? Pls test.

Comment (komalsukhani) [reply gateway/mw_request_signing.go]: Will check

**Comment by komalsukhani** [comment]
@buger I have made changes to handle case of url-rewrite and added few more checks on input.

**Pull request merged**